### DISCUSSION TITLE
Techniques for Debugging XPath Feeds?

### DISCUSSION THREAD

**Discussion by mikelococo** [FreshRSS/FreshRSS#6011] [Q&A] [answered]
### Summary

Are there any generally useful techniques for debugging HTML+XPath feeds that are misbehaving? My first XPath feed setup went incredibly smoothly and I'm really liking extracting news items from sites that don't expose proper RSS. But I'm trying to replicate my success with a second feed and feeling like I'm getting essentially no useful feedback from FreshRSS about what is going wrong.

### General Techniques?

Helpful information I'd expect it to be possible to collect, but which I haven't found a way to see might include...
- Getting FreshRSS to show me the unprocessed HTML that was scraped, so I can verify that the site isn't showing some cloudflare-captcha, anti-scraping, or otherwise busted page content instead of the content I see in my browser.
- Seeing some output related what XPath queries were run on the HTML content and some indication of what the results were (error, no match, or something else).
- Really any other context it's possible to extract about what went right or wrong about the XPath parsing.

### My Broken Feed

In addition to the general technique questions above, I'll provide some concrete context on the Feed I'm trying to put together...

1. I'm trying to extract the items in the "News" carousel at https://foundryvtt.com/.
2. The relevant bits of HTML look roughly like this (I've simplified it a bit, but not much... also... note that the "carousel-item" entries without the "active" class have a trailing space... I sort of wonder if that's throwing the parser off):
   ```
   <nav id="news" class="carousel">
     <div class="carousel-figures">
       <figure class="carousel-item active">
         <some-content />
       </figure>
       <figure class="carousel-item ">
         <some-content />
       </figure>
       <figure class="carousel-item ">
         <some-content />
       </figure>
     </div>
   </nav>
   ```
3. I've tried a variety of XPaths for finding news items, but the simplest one I think should return something useful is `//figure[@class="carousel-item"]`
4. This is currently returning an empty feed with no messages in the Logs GUI when I click refresh feed, and no log messages in the FreshRSS docker logs either (other than the `GET /i/` and `POST /i/` logs that I think Apache generate when I interact with the FreshRSS UI).
5. At various other times (and I believe with this exact same XPath), I've gotten ` HTML+XPath Web scraping failed for [https://foundryvtt.com] ` errors in the logs UI and those are accompanied by docker logs like this :
   ```
   Dec 31 22:49:46 srvr freshrss-pod-freshrss[3833409]: FreshRSS[40]: FreshRSS uses cache for https://foundryvtt.com
   Dec 31 22:49:46 srvr freshrss-pod-freshrss[3833409]: FreshRSS[40]: [mike] [Sun, 31 Dec 2023 17:49:46 -0500] [warning] --- HTML+XPath Web scraping failed for [https://foundryvtt.com]
   ```
   Or if I clear the cache for the feed I can see the GET log instead of the cached log:
   ```
   Dec 31 23:05:41 srvr freshrss-pod-freshrss[3833409]: FreshRSS[46]: FreshRSS GET html https://foundryvtt.com
   Dec 31 23:05:42 srvr freshrss-pod-freshrss[3833409]: FreshRSS[46]: [mike] [Sun, 31 Dec 2023 18:05:42 -0500] [warning] --- HTML+XPath Web scraping failed for [https://foundryvtt.com]
   ```
6. I have once managed to get the single active news item with an XPath of `//figure[@class="carousel-item active"]` though things broke when I removed the additional "active" class to try to pick up the rest of the entries. And I can't even replicate that now. If I try that now I get an empty feed with no logs.

It all feels very much like maybe the site content is shifting around underneath me, but I just feel like I have no tools to try to debug with. Any suggestions on the specific case or general techniques are much appreciated.

**Comment by quella** [comment, marked as answer]
I started to have a look at this using Google's Inspect Element option and can see that the webpage is dynamic and there are 10 news items that shift and "reload" the items in the list of news.  I can see this actively happening so I do not think this would work for scraping; just my opinion.   

When I started to look at this (as I'm also learning scraping), I was getting the below items until I saw that the items did not have content (when this happens I often just use the article title as well.  I use Google's Element Inspect under the Development (view) as I can paste back what I think is what I'm looking for and I can see if I get back the right context.  I also have hit a few blocks on sites that I think implement some form of anti-scraping, but have not had a good way to confirm if this is the case or not.  I posted about that a few days ago.  

I even said in that post that it would be good to have a place where people could upload their successful scraping and share them with others that may desire the same and not have the time or ability to scrape.  It is such a great technology!  Best of luck.

Title: | //title
Xpath to Items: | //*[@id="news"]/div/figure
Item Title (Relative) | //*[@id="news"]/div/figure/descendant::a/@title
Item Content (Relative) | //*[@id="news"]/div/figure/descendant::a/@title
Item Link URL (Relative) | //*[@id="news"]/div/figure/descendant::a/img/@src

I stopped with the above when I saw the rotation dynamically in the page.  Not sure how a RSS feed would handle that as they would not know where to end or start the next time.

**Comment by mikelococo** [comment]
> the webpage is dynamic and there are 10 news items that shift and "reload" the items in the list of news. I can see this actively happening so I do not think this would work for scraping

Thanks so much for your response. I'm inclined to believe this. I was also using a developer-tools inspect view (in Firefox) but missed the dynamism. I think it probably would have stuck out like a sore thumb if I could have seen the content as scraped by FreshRSS though.

I'm happy to give up on this particular feed, the critical product news that lands here gets mirrored at other sources that are more easily subscribable. I think the main takeaway for me though is that some additional visibility in FreshRSS's UI or logs would provide a lot of value here. It takes a large amount of very finicky mental modelling to speculate about what might be going on behind the scenes in any given case and a bit more observability would go a long way.

**Comment by quella** [comment]
I would agree that some additional debugging would be nice to have in the FreshRSS tool and logs.  I'm glad to have the option to scrape and at this time have more been plotting though it with the data provided.  I know a few others have posted data on the ability to scrape and doing it in FreshRSS, but these often use examples of very easy sites (bookstore) or assume you have a load of experience with XPATH and over complicate things.  Really a magic ability that seems like it takes practice.  When it works, it is also rather magical.  :)
